dot/rpc/modules/chain_test.go

```diff
@@ -15,3 +15,473 @@
 // along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
 
 package modules
+
+import (
+	"errors"
+	"math/big"
+	"net/http"
+	"testing"
+
+	apimocks "github.com/ChainSafe/gossamer/dot/rpc/modules/mocks"
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestChainModule_GetBlock(t *testing.T) {
+	mockedHash := common.NewHash([]byte{0x01, 0x02})
+	emptyBlock := types.NewEmptyBlock()
+
+	mockBlockAPI := new(apimocks.BlockAPI)
+	mockBlockAPI.On("GetBlockByHash", mock.AnythingOfType("common.Hash")).Return(&emptyBlock, nil)
+	mockBlockAPI.On("BestBlockHash").Return(mockedHash, nil)
+
+	mockBlockAPIGetHashErr := new(apimocks.BlockAPI)
+	mockBlockAPIGetHashErr.On("GetBlockByHash", mock.AnythingOfType("common.Hash")).Return(nil, errors.New("GetJustification error"))
+
+	bodyBlock := types.NewEmptyBlock()
+	bodyBlock.Body = types.BytesArrayToExtrinsics([][]byte{{1}})
+	mockBlockAPIWithBody := new(apimocks.BlockAPI)
+	mockBlockAPIWithBody.On("GetBlockByHash", mock.AnythingOfType("common.Hash")).Return(&bodyBlock, nil)
+
+	chainModule := NewChainModule(mockBlockAPI)
+	res := ChainBlockResponse{}
+	type fields struct {
+		blockAPI BlockAPI
+	}
+	type args struct {
+		r   *http.Request
+		req *ChainHashRequest
+		res *ChainBlockResponse
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "GetBlock OK",
+			fields: fields{
+				chainModule.blockAPI,
+			},
+			args: args{
+				r:   nil,
+				req: &ChainHashRequest{},
+				res: &res,
+			},
+			wantErr: false,
+		},
+		{
+			name: "GetBlockByHash Err",
+			fields: fields{
+				mockBlockAPIGetHashErr,
```

Comment:
You should define these mocks in the actual subtest below. Instead you can have parameters to configure your mock, here I guess the expected hash and the error to return. Otherwise they'll conflict with each other, especially more once you start running parallel subtests.

Reply:
So I'm running the tests in parallel now and haven't gotten conflicts, is it still best to make this change?

Reply:
Try

```sh
CGO_ENABLED=1 go test -race -run ^TestAuthorModule_HasSessionKey_WhenScaleDataEmptyOrNil$ github.com/ChainSafe/gossamer/dot/rpc/modules
```

To see if it finds any races on the mocks? If not leave it as is and my apologies!

Reply:
Okay for that test specifically there weren't issues, but when I ran it on the whole pkg I had conflicts. Will make these changes!

Reply:
> Okay for that test specifically there weren't issues

Are you running the subtests below with `t.Parallel()` though? If the race detector doesn't throw up on the single test, it means all the subtests don't data race on shared variables (like the mock). That might be true as well since the mock might just be read/executed and not written to.

> but when I ran it on the whole pkg I had conflicts

That could be due to a million different things, don't invest too much time in that either.
Although fixing the data races on tests not running in parallel would be nice, so we can sometime soon setup the CI to use the race detector to detect newer data races introduced.

Reply:
Yeah I have t.Parallel() in both the test and subtests. So do you think then I should make these changes, and if there are still conflicts after when I run the pkg tests with -race move on for now?

Reply:
If you are rewriting all the tests in the package, you might want to fix the data races **before** adding `t.Parallel()` to tests.

If you find data races in existing tests running without parallelism, then it might be some critical data race in the production code.

Now when you  all your tests in parallel (with `t.Parallel()`), it will likely reveal additional data races, which are mostly test only data races (e.g. writing to a shared test global variable).

Let me know if you need help figuring it out, we can pair program tomorrow for ~30 min or so to fix them for the tests of that package.